### PR TITLE
implement BigFloat negation and abs without ccall

### PR DESCRIPTION
**Pull request by nsajko**

### CODE REVIEW

File: src/implementations/BigFloat.jl [hunk 1]
```diff
@@ -79,28 +79,14 @@ function operate_to!(output::BigFloat, ::typeof(-), a::BigFloat, b::BigFloat)
...
 function operate!(::typeof(-), x::BigFloat)
-    ccall(
-        (:mpfr_neg, :libmpfr),
-        Int32,
-        (Ref{BigFloat}, Ref{BigFloat}, Base.MPFR.MPFRRoundingMode),
-        x,
-        x,
-        Base.MPFR.ROUNDING_MODE[],
-    )
+    x.sign = -x.sign
```
Comment (odow) [inline src/implementations/BigFloat.jl:82]: Is the `.sign` field public in `BigFloat`? I couldn't find it documented anywhere. We have to be careful about modifying internal parameters of structs, even if that's what the `ccall` ends up doing.

Comment (nsajko) [reply src/implementations/BigFloat.jl:82]: It's not documented, but I think the same holds for `ccall`-ing libmpfr, so I guess you're already in that territory?

Comment (odow) [reply src/implementations/BigFloat.jl:82]: I guess `libmpfr` is a library with public methods. But the Julia `struct BigFloat` could rename the fields and still be non-breaking with respect to the C API (it just cares about the order, size, and offset of the struct fields, not their name in Julia).

But this seems very low risk, so I'm okay with it. Let's see what @blegat thinks.

Comment (nsajko) [reply src/implementations/BigFloat.jl:82]: I guess Julia could technically move to a pure-Julia `BigFloat` implementation that wouldn't use MPFR at all. That would hypothetically break all the `ccall`s.

Comment (nsajko) [reply src/implementations/BigFloat.jl:82]: So I'd say both options are undocumented and unsupported by Julia, at least technically.

Comment (odow) [reply src/implementations/BigFloat.jl:82]: I'm going to merge. I'll take the blame if this goes badly in future :smile:

Comment (nsajko) [reply src/implementations/BigFloat.jl:82]: TBH there's definitely a chance for Julia `Base` breaking MA, a bit more so with this change. I think the only long term solution is exposing a MA-like interface as a public API of Julia.

Comment (odow) [reply src/implementations/BigFloat.jl:82]: Yeah, but we do keep an eye on testing upcoming versions of Julia. And there are plenty of other ways that Base can break MA by changing SparseArrays or LinearAlgebra, so we're already living dangerously.

**Pull request merged**